### DISCUSSION TITLE
Kong Go plugin execution model

### DISCUSSION THREAD

**Discussion by BJL** [Kong/kong#10943] [Help] [answered]
Hi,

As far as I understand, a GO plugin for Kong is a simple GO process listening on a socket (unix, tcp,...) for requests.

I was wondering if I have to manage multithreading in this GO process (go routines) in order to avoid contention under high traffic.
As Kong is based on Nginx/open-resty which has a strong multi process / multi thread model, I imagine that, at some point, it may saturate a single mono GO process.

Any information about that ?

Regards,

Ben

**Comment by ms2008** [comment, marked as answer]
This is a brief technical implementation, for folks interested following through the code of Kong's [go-plugins](https://github.com/Kong/go-plugins) to see where the concurrency is actually implemented.

The relevant source code for the serving part of go-plugins is in https://github.com/Kong/go-pdk/blob/master/server/pbserver.go

`StartServer` which calls `Accept` in a loop. `Accept` is similar to the socket accept syscall, creating a new connection whenever a new client is accepted; the connection is then used for interacting with the client. This connection is `type UnixConn` in the Go's source `net/unixsock.go` file, a type with server state for each client connection. 

`StartServer` calls `servePb` as:

```
go servePb(conn, rh)
```

Wherein lies the concurrency. From this point on, a separate goroutine handles this connection. This is why multiple goroutines could be executing user-specified handlers (including a single handler) at any given time.

P.S. Creating a goroutine for each client connection is a common model for Go's concurrent programming.

**Reply by BJL** [reply]
thx !
